File: app/src/main/java/de/nils_beyer/android/Vertretungen/mainActivity/OverviewSectionsAdapter.java
package de.nils_beyer.android.Vertretungen.mainActivity;

import android.content.Context;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.view.ViewGroup;

import java.util.Date;

import de.nils_beyer.android.Vertretungen.account.AccountManager;
import de.nils_beyer.android.Vertretungen.data.GroupCollection;
import de.nils_beyer.android.Vertretungen.storage.StudentStorage;
import de.nils_beyer.android.Vertretungen.util.DateParser;
import de.nils_beyer.android.Vertretungen.R;


class OverviewSectionsAdapter extends FragmentPagerAdapter {

    private OverviewFragment f0;
    private OverviewFragment f1;
    private Context context;
    private DownloadingActivity downloadingActivity;
    private GroupCollection g1;
    private GroupCollection g2;


    OverviewSectionsAdapter(Context c, FragmentManager fm, DownloadingActivity _downloadingActivity, GroupCollection g1, GroupCollection g2) {
        super(fm);
        context = c;
        downloadingActivity = _downloadingActivity;
        this.g1 = g1;
        this.g2 = g2;
    }

    @Override
    public Fragment getItem(int position) {

        switch (position) {
            case 0:
                f0 = OverviewFragment.getIntance(context, g1);
                return f0;
            case 1:
                f1 = OverviewFragment.getIntance(context, g2);
                return f1;
        }
        return null;
    }

    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        Fragment of = (Fragment) super.instantiateItem(container, position);
        switch (position) {
            case 0:
                f0 = (OverviewFragment) of;
                break;
            case 1:
                f1 = (OverviewFragment) of;
                break;
        }

        checkIfDownloading();

        return of;
    }

    @Override
    public int getCount() {
        // Heute und Morgen => 2
        return 2;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        final Date dateToday = StudentStorage.getDateToday(context);
        final Date dateTomorrow = StudentStorage.getDateTomorrow(context);

        switch (position) {
            case 0:
                if (dateToday == null)
                    return context.getString(R.string.today);
                else
                    return DateParser.parseDateToShortString(context, dateToday);

            case 1:
                if (dateTomorrow == null)
                    return context.getString(R.string.tomorrow);
                else
                    return DateParser.parseDateToShortString(context, dateTomorrow);
        }
        return null;
    }

    @Override
    public void notifyDataSetChanged() {
        // Update Fragments
        if (f0 != null) {
            f0.updateData(g1);
        }

        if (f1 != null) {
            f1.updateData(g2);
        }

        super.notifyDataSetChanged();
    }

    public void update(GroupCollection g1, GroupCollection g2) {
        this.g1 = g1;
        this.g2 = g2;
        notifyDataSetChanged();
    }

    public void hideDownloading() {
        // Hide Spinning Button
        if (f0 != null) {
            f0.resetSwipeRefreshLayout();
        }

        if (f1 != null) {
            f1.resetSwipeRefreshLayout();
        }
    }

    public void showDownloading() {
        // Enable Spinning Button
        if (f0 != null) {
            f0.showSwipeRefreshLayout();
        }

        if (f1 != null) {
            f1.showSwipeRefreshLayout();
        }
    }

    private void checkIfDownloading() {
        if (downloadingActivity.isDownloading()) {
            showDownloading();
        } else {
            hideDownloading();
        }
    }

    public interface DownloadingActivity {
        boolean isDownloading();
    }
}

File: app/src/main/java/de/nils_beyer/android/Vertretungen/mainActivity/OverviewSectionsAdapter.kt
package de.nils_beyer.android.Vertretungen.mainActivity


import android.content.Context
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import android.view.ViewGroup
import de.nils_beyer.android.Vertretungen.R


import de.nils_beyer.android.Vertretungen.util.DateParser
import de.nils_beyer.android.Vertretungen.account.Dataset
import de.nils_beyer.android.Vertretungen.account.Account


internal class OverviewSectionsAdapter(private val context: Context, fm: FragmentManager, private val downloadingActivity: DownloadingActivity, private var account : Account<out Dataset>) : FragmentStatePagerAdapter(fm) {

    //private var fragments = arrayOfNulls<Fragment>(count);
    private var datasets = account.getAvailableDatasets().map { it.getData(context) }

    override fun getItem(position: Int): Fragment {
        if (position == 0) {
            return EventListFragment();
        } else {
            return OverviewFragment.getIntance(context, datasets[position - 1], position)
        }
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val of = super.instantiateItem(container, position)
        if (of is OverviewFragment) {
            checkIfDownloading()
        }
        return of
    }

    override fun getCount(): Int = account.getAvailableDatasets().size + 1 // plus one for the calendar view

    override fun getPageTitle(position: Int): CharSequence? {
        if (position == 0) {
            // calendar view
            return context.getString(R.string.tab_events_name)
        } else {
            // substitutions view
            val date = datasets[position - 1].date

            if (date == null) {
                return String.format(context.getString(R.string.tab_substitutions_name),
                        position - 1)
            } else {
                return DateParser.parseDateToShortString(context, date)
            }
        }
    }

    override fun getItemPosition(pFragment: Any): Int {
        if (pFragment is OverviewFragment) {
            val position = pFragment.position;
            pFragment.updateData(datasets[position - 1])
            return position
        } else if (pFragment is EventListFragment) {
            pFragment.update()
            return 0
        }
        return FragmentStatePagerAdapter.POSITION_NONE
    }


    /**
     * Apply the Adapter to a new account and enforce a complete reload of the fragments
     * @param account the account which data should now be displayed.
     */
    fun update(account: Account<*>) {
        this.account = account;
        datasets = account.getAvailableDatasets().map { it.getData(context) }
        // When the new account has the same amount of tabs as the previous one,
        // we can update each tab so that the fragment stays in place.
        /*if (fragments.size == count) {

        } else {
            // if not, we have to create a new array of fragments where we store
            // the fragments
            val fragmentsCopy = arrayOfNulls<Fragment>(count)
            fragments.forEachIndexed { index, overviewFragment ->
                if (overviewFragment is OverviewFragment && index < account.getAvailableDatasets().size)
                    fragmentsCopy[index] = overviewFragment
            }
            fragments = fragmentsCopy
        }*/

        notifyDataSetChanged()
    }

    fun getDataset(position : Int) : Dataset? {
        if (position - 1 < 0 || position - 1 >= account.getAvailableDatasets().size) {
            //throw RuntimeException("no groupcollection at index ${position} available.")
            return null;
        }
        return account.getAvailableDatasets()[position - 1]
    }

    fun hideDownloading() {
        // Hide Spinning Button
        /*fragments.forEach {
            if (it is OverviewFragment)
                it?.resetSwipeRefreshLayout()
        }*/
    }

    fun showDownloading() {
        // Enable Spinning Button
        /*fragments.forEach {
            if (it is OverviewFragment)
                it?.showSwipeRefreshLayout()
        }*/
    }

    private fun checkIfDownloading() {
        if (downloadingActivity.isDownloading) {
            showDownloading()
        } else {
            hideDownloading()
        }
    }

    interface DownloadingActivity {
        val isDownloading: Boolean
    }
}